<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="PersonalisIO">
<title>Loading Personalis data • PersonalisIO</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Loading Personalis data">
<meta property="og:description" content="PersonalisIO">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">PersonalisIO</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/PersonalisIO.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Loading Personalis data</h1>
            
      
      
      <div class="d-none name"><code>PersonalisIO.Rmd</code></div>
    </div>

    
    
<p>This vignette demonstrates how to use the {PersonalisIO} package to
load Personalis data into a {MultiAssayExperiment} object.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://Boehringer-Ingelheim.github.io/Personalis-io" class="external-link">PersonalisIO</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/" class="external-link">purrr</a></span><span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://conflicted.r-lib.org/" class="external-link">conflicted</a></span><span class="op">)</span></span>
<span>  <span class="fu"><a href="https://conflicted.r-lib.org/reference/conflicts_prefer.html" class="external-link">conflicts_prefer</a></span><span class="op">(</span><span class="fu">base</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/r/base/sets.html" class="external-link">setdiff</a></span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">[conflicted]</span> Will prefer <span style="color: #0000BB; font-weight: bold;">base</span>::setdiff over any</span></span>
<span><span class="co">#&gt; other package.</span></span></code></pre></div>
<p>Personalis ship their data in a (more or less) standardized directory
structure within one folder for each patient. The PersonalisIO loader
functions traverse this directory structure and extract the different
modalities. To load data from multiple patients into a single object, we
need a list of patient directories:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">path</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">"PERSONALIS_PATH"</span><span class="op">)</span></span>
<span><span class="va">sample_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">path</span>, full.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/pkg/purrr/man/keep.html" class="external-link">keep</a></span><span class="op">(</span><span class="va">dir.exists</span><span class="op">)</span></span>
<span><span class="va">sample_list</span></span>
<span><span class="co">#&gt; [1] "/tmp/personalis_example_data/sample1"</span></span>
<span><span class="co">#&gt; [2] "/tmp/personalis_example_data/sample2"</span></span>
<span><span class="co">#&gt; [3] "/tmp/personalis_example_data/sample3"</span></span>
<span><span class="co">#&gt; [4] "/tmp/personalis_example_data/sample4"</span></span>
<span><span class="co">#&gt; [5] "/tmp/personalis_example_data/sample5"</span></span></code></pre></div>
<div class="section level2">
<h2 id="reading-single-modalities-into-a-summarizedexperiment">Reading single modalities into a SummarizedExperiment<a class="anchor" aria-label="anchor" href="#reading-single-modalities-into-a-summarizedexperiment"></a>
</h2>
<p>We can now read a single modality into a {SummarizedExperiment}. For
instance, to read gene expression data only.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">personalis_gex</span> <span class="op">&lt;-</span> <span class="fu">PersonalisIO</span><span class="fu">::</span><span class="fu"><a href="../reference/read_personalis_gene_expression.html">read_personalis_gene_expression</a></span><span class="op">(</span><span class="va">sample_list</span><span class="op">)</span></span></code></pre></div>
<p>By inspecting the object, we can see, this imported gene expression
data from 5 patients and ~23k genes. There are three different assays:
<code>counts</code>, <code>tpm</code> and <code>normCounts</code>:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">personalis_gex</span></span>
<span><span class="co">#&gt; class: SummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 22956 5 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(3): counts tpm normCounts</span></span>
<span><span class="co">#&gt; rownames(22956): 1 503538 ... 23140 26009</span></span>
<span><span class="co">#&gt; rowData names(1): Gene Symbol</span></span>
<span><span class="co">#&gt; colnames(5): sample1 sample2 sample3 sample4 sample5</span></span>
<span><span class="co">#&gt; colData names(0):</span></span></code></pre></div>
<p>The following functions to load individual modalities exist, and they
all return a {SummarizedExperiment}:</p>
<pre><code><span><span class="va">read_personalis_gene_expression</span></span>
<span><span class="va">read_personalis_small_variant_reports</span></span>
<span><span class="va">read_personalis_cnv_reports</span></span>
<span><span class="va">read_personalis_hla_reports</span></span>
<span><span class="va">read_personalis_tcr_reports</span></span>
<span><span class="va">read_personalis_msi_reports</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="reading-multiple-modalities-into-a-multiassayexperiment">Reading multiple modalities into a MultiAssayExperiment<a class="anchor" aria-label="anchor" href="#reading-multiple-modalities-into-a-multiassayexperiment"></a>
</h2>
<p>To import all modalities at once, we can use the
<code>read_personalis</code> function. Not all modalities may be present
for all samples. Some may not be present in any sample of the dataset.
This is shown as warnings during the import process. It is your
repsonsibility to check the warnings carefully to make sure that all
data that you expect to be there is successfully imported.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mae</span> <span class="op">&lt;-</span> <span class="fu">PersonalisIO</span><span class="fu">::</span><span class="fu"><a href="../reference/read_personalis.html">read_personalis</a></span><span class="op">(</span><span class="va">sample_list</span><span class="op">)</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading gene expression...</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading DNA small variant data...</span></span>
<span><span class="co">#&gt; Reading DNA small variant data (somatic)...</span></span>
<span><span class="co">#&gt; Reading DNA small variant data (tumor)...</span></span>
<span><span class="co">#&gt; Reading DNA small variant data (normal)...</span></span>
<span><span class="co">#&gt; ⚠️ No gene DNA small variant (normal) found for sample1. Sample ignored.</span></span>
<span><span class="co">#&gt; ⚠️ No gene DNA small variant (normal) found for sample2. Sample ignored.</span></span>
<span><span class="co">#&gt; ⚠️ No gene DNA small variant (normal) found for sample3. Sample ignored.</span></span>
<span><span class="co">#&gt; ⚠️ No gene DNA small variant (normal) found for sample4. Sample ignored.</span></span>
<span><span class="co">#&gt; ⚠️ No gene DNA small variant (normal) found for sample5. Sample ignored.</span></span>
<span><span class="co">#&gt; Warning in warn_missing_samples(sample_list, sample_paths, description): 5 DNA</span></span>
<span><span class="co">#&gt; small variant (normal) sample(s) were not found. Please check the log for</span></span>
<span><span class="co">#&gt; details.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading RNA small variant data...</span></span>
<span><span class="co">#&gt; Reading RNA small variant data (somatic)...</span></span>
<span><span class="co">#&gt; Reading RNA small variant data (tumor)...</span></span>
<span><span class="co">#&gt; Reading RNA small variant data (normal)...</span></span>
<span><span class="co">#&gt; ⚠️ No gene RNA small variant (normal) found for sample1. Sample ignored.</span></span>
<span><span class="co">#&gt; ⚠️ No gene RNA small variant (normal) found for sample2. Sample ignored.</span></span>
<span><span class="co">#&gt; ⚠️ No gene RNA small variant (normal) found for sample3. Sample ignored.</span></span>
<span><span class="co">#&gt; ⚠️ No gene RNA small variant (normal) found for sample4. Sample ignored.</span></span>
<span><span class="co">#&gt; ⚠️ No gene RNA small variant (normal) found for sample5. Sample ignored.</span></span>
<span><span class="co">#&gt; Warning in warn_missing_samples(sample_list, sample_paths, description): 5 RNA</span></span>
<span><span class="co">#&gt; small variant (normal) sample(s) were not found. Please check the log for</span></span>
<span><span class="co">#&gt; details.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading microsatellite stability (MSI) data...</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading copy number variation (CNV) data...</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading HLA types...</span></span>
<span><span class="co">#&gt; ⚠️ No gene HLA found for sample1. Sample ignored.</span></span>
<span><span class="co">#&gt; ⚠️ No gene HLA found for sample2. Sample ignored.</span></span>
<span><span class="co">#&gt; ⚠️ No gene HLA found for sample3. Sample ignored.</span></span>
<span><span class="co">#&gt; ⚠️ No gene HLA found for sample4. Sample ignored.</span></span>
<span><span class="co">#&gt; ⚠️ No gene HLA found for sample5. Sample ignored.</span></span>
<span><span class="co">#&gt; Warning in warn_missing_samples(sample_list, sample_paths, description): 5 HLA</span></span>
<span><span class="co">#&gt; sample(s) were not found. Please check the log for details.</span></span>
<span><span class="co">#&gt; &gt;&gt;&gt; Reading TCR clone reports...</span></span></code></pre></div>
<p>Here’s an overview of the {MultiAssayExperiment} object with all
available modalities:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mae</span></span>
<span><span class="co">#&gt; A MultiAssayExperiment object of 8 listed</span></span>
<span><span class="co">#&gt;  experiments with user-defined names and respective classes.</span></span>
<span><span class="co">#&gt;  Containing an ExperimentList class object of length 8:</span></span>
<span><span class="co">#&gt;  [1] gene_expression: SummarizedExperiment with 22956 rows and 5 columns</span></span>
<span><span class="co">#&gt;  [2] dna_small_variants_somatic: SummarizedExperiment with 6097 rows and 5 columns</span></span>
<span><span class="co">#&gt;  [3] dna_small_variants_tumor: SummarizedExperiment with 376654 rows and 5 columns</span></span>
<span><span class="co">#&gt;  [4] rna_small_variants_somatic: SummarizedExperiment with 5302 rows and 5 columns</span></span>
<span><span class="co">#&gt;  [5] rna_small_variants_tumor: SummarizedExperiment with 201347 rows and 5 columns</span></span>
<span><span class="co">#&gt;  [6] msi: SummarizedExperiment with 5 rows and 5 columns</span></span>
<span><span class="co">#&gt;  [7] cnv: SummarizedExperiment with 145 rows and 5 columns</span></span>
<span><span class="co">#&gt;  [8] tcr: SummarizedExperiment with 2 rows and 5 columns</span></span>
<span><span class="co">#&gt; Functionality:</span></span>
<span><span class="co">#&gt;  experiments() - obtain the ExperimentList instance</span></span>
<span><span class="co">#&gt;  colData() - the primary/phenotype DataFrame</span></span>
<span><span class="co">#&gt;  sampleMap() - the sample coordination DataFrame</span></span>
<span><span class="co">#&gt;  `$`, `[`, `[[` - extract colData columns, subset, or experiment</span></span>
<span><span class="co">#&gt;  *Format() - convert into a long or wide DataFrame</span></span>
<span><span class="co">#&gt;  assays() - convert ExperimentList to a SimpleList of matrices</span></span>
<span><span class="co">#&gt;  exportClass() - save data to flat files</span></span></code></pre></div>
<p>We can retrieve an individual modality by subsetting with double
brackets:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mae</span><span class="op">[[</span><span class="st">"gene_expression"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; class: SummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 22956 5 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(3): counts tpm normCounts</span></span>
<span><span class="co">#&gt; rownames(22956): 1 503538 ... 23140 26009</span></span>
<span><span class="co">#&gt; rowData names(1): Gene Symbol</span></span>
<span><span class="co">#&gt; colnames(5): sample1 sample2 sample3 sample4 sample5</span></span>
<span><span class="co">#&gt; colData names(0):</span></span></code></pre></div>
<p>For more information about advanced subsetting, please refer to the
<a href="https://bioconductor.org/packages/release/bioc/vignettes/MultiAssayExperiment/inst/doc/MultiAssayExperiment.html#integrated-subsetting-across-experiments" class="external-link">MultiAssayExperiment
vignette</a>.</p>
</div>
<div class="section level2">
<h2 id="summary-statistics">Summary statistics<a class="anchor" aria-label="anchor" href="#summary-statistics"></a>
</h2>
<p>For some modalities, patient-level summary statistics are provided by
Personalis in their HTML reports. These are gathered automatically by
the respective reader function, and stored in <code>colData</code>. For
instance, tumor mutational burden is available from the
<code>dna_small_variants_somatic</code> modality:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="st">"dna_small_variants_somatic"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "Somatic variants (SNVs)"                        </span></span>
<span><span class="co">#&gt;  [2] "Somatic variants (Indels)"                      </span></span>
<span><span class="co">#&gt;  [3] "Somatic variants (Total)"                       </span></span>
<span><span class="co">#&gt;  [4] "Somatic variants per Mb (SNVs)"                 </span></span>
<span><span class="co">#&gt;  [5] "Somatic variants per Mb (Indels)"               </span></span>
<span><span class="co">#&gt;  [6] "Somatic variants per Mb (Total)"                </span></span>
<span><span class="co">#&gt;  [7] "Ti/Tv ratio (SNVs)"                             </span></span>
<span><span class="co">#&gt;  [8] "Ti/Tv ratio (Indels)"                           </span></span>
<span><span class="co">#&gt;  [9] "Ti/Tv ratio (Total)"                            </span></span>
<span><span class="co">#&gt; [10] "Non-synonymous Somatic Variants (SNVs)"         </span></span>
<span><span class="co">#&gt; [11] "Non-synonymous Somatic Variants (Indels)"       </span></span>
<span><span class="co">#&gt; [12] "Non-synonymous Somatic Variants (Total)"        </span></span>
<span><span class="co">#&gt; [13] "Non-synonymous Somatic Variants per Mb (SNVs)"  </span></span>
<span><span class="co">#&gt; [14] "Non-synonymous Somatic Variants per Mb (Indels)"</span></span>
<span><span class="co">#&gt; [15] "Non-synonymous Somatic Variants per Mb (Total)"</span></span></code></pre></div>
<p>Let’s take a closer look at the different mutation rates</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">mae</span><span class="op">[[</span><span class="st">"dna_small_variants_somatic"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/tibble/man/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span>rownames <span class="op">=</span> <span class="st">"sample"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">sample</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">starts_with</a></span><span class="op">(</span><span class="st">"Somatic.variants.per.Mb"</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="7%">
<col width="29%">
<col width="31%">
<col width="30%">
</colgroup>
<thead><tr class="header">
<th align="left">sample</th>
<th align="right">Somatic.variants.per.Mb..SNVs.</th>
<th align="right">Somatic.variants.per.Mb..Indels.</th>
<th align="right">Somatic.variants.per.Mb..Total.</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">sample1</td>
<td align="right">18.82</td>
<td align="right">0.70</td>
<td align="right">19.52</td>
</tr>
<tr class="even">
<td align="left">sample2</td>
<td align="right">13.14</td>
<td align="right">0.39</td>
<td align="right">13.53</td>
</tr>
<tr class="odd">
<td align="left">sample3</td>
<td align="right">12.95</td>
<td align="right">0.58</td>
<td align="right">13.53</td>
</tr>
<tr class="even">
<td align="left">sample4</td>
<td align="right">10.68</td>
<td align="right">0.48</td>
<td align="right">11.16</td>
</tr>
<tr class="odd">
<td align="left">sample5</td>
<td align="right">26.62</td>
<td align="right">1.05</td>
<td align="right">27.67</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="working-with-bumpymatrices">Working with BumpyMatrices<a class="anchor" aria-label="anchor" href="#working-with-bumpymatrices"></a>
</h2>
<p>Not all Personalis data is strictly rectangular like a gene
expression matrix and nicely fits into a SummarizedExperiment. For
instance, TCR clones or small variant reports are a long-form data frame
with many entries for each patient. The {BumpyMatrix} package comes to
the rescue. It allows to fit long-form data frames into a
{SummarizedExperiment} by building a matrix that is rectangular in the
first two dimensions, but is ragged (i.e. arbitrary length) in the third
dimension.</p>
<p>In practice, bumpy matrices are a bit cumbersome to work with. They
are to be seen as an intermediate container that allows to store
particular kinds of data in the MultiAssayExperiment object and (after
subsetting to the patients of interest) can be converted back to a data
frame using <code><a href="https://rdrr.io/pkg/BumpyMatrix/man/unsplitAsDataFrame.html" class="external-link">BumpyMatrix::unsplitAsDataFrame</a></code>.</p>
<div class="section level3">
<h3 id="example-tcr-data-stored-as-bumpymatrix">Example: TCR data stored as BumpyMatrix<a class="anchor" aria-label="anchor" href="#example-tcr-data-stored-as-bumpymatrix"></a>
</h3>
<p>The <code>tcr</code> assay of our <code>mae</code> object is such a
BumpyMatrix. Here, the colnames correspond to our patient identifiers,
and the row names to the TCR chain (TRA = alpha, TRB = beta). Each cell
of the matrix is a data frame with all clones for the respective patient
and chain.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tcr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">mae</span>, <span class="st">"tcr"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">tcr</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "BumpyDataFrameMatrix"</span></span>
<span><span class="co">#&gt; attr(,"package")</span></span>
<span><span class="co">#&gt; [1] "BumpyMatrix"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">tcr</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "sample1" "sample2" "sample3" "sample4" "sample5"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">tcr</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "TRA" "TRB"</span></span></code></pre></div>
<p>It is possible to retrieve this data frame for a single chain and
patient:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tcr</span><span class="op">[</span><span class="st">"TRA"</span>, <span class="st">"sample1"</span><span class="op">]</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/tibble/man/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">Clonal.Sequence</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 90 × 10</span></span></span>
<span><span class="co">#&gt;    group group_name Clone.Count Clone.Frequency Top.V.Hits       Top.J.Hits</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;int&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span>     1 <span style="color: #BB0000;">NA</span>                 193          0.055<span style="text-decoration: underline;">4</span> TRAV8-6, TRAV8-4 TRAJ5     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span>     1 <span style="color: #BB0000;">NA</span>                 193          0.055<span style="text-decoration: underline;">4</span> TRAV26-1         TRAJ44    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span>     1 <span style="color: #BB0000;">NA</span>                 158          0.045<span style="text-decoration: underline;">3</span> TRAV22           TRAJ28    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span>     1 <span style="color: #BB0000;">NA</span>                 138          0.039<span style="text-decoration: underline;">6</span> TRAV6            TRAJ41    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span>     1 <span style="color: #BB0000;">NA</span>                 137          0.039<span style="text-decoration: underline;">3</span> TRDV1            TRAJ23    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span>     1 <span style="color: #BB0000;">NA</span>                 136          0.039<span style="text-decoration: underline;">0</span> TRAV17           TRAJ12    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span>     1 <span style="color: #BB0000;">NA</span>                 130          0.037<span style="text-decoration: underline;">3</span> TRAV13-1         TRAJ12    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span>     1 <span style="color: #BB0000;">NA</span>                 129          0.037<span style="text-decoration: underline;">0</span> TRAV19           TRAJ13    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span>     1 <span style="color: #BB0000;">NA</span>                 121          0.034<span style="text-decoration: underline;">7</span> TRAV13-1         TRAJ57    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span>     1 <span style="color: #BB0000;">NA</span>                 118          0.033<span style="text-decoration: underline;">9</span> TRAV21           TRAJ12    </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 80 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 4 more variables: CDR3.Amino.Acid.Sequence &lt;chr&gt;, locus &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   sample &lt;chr&gt;, Top.D.Hits &lt;chr&gt;</span></span></span></code></pre></div>
<p>We can also slice the matrix for a single variable, e.g. we can use
this to retrieve only the CDR3 amino acid sequence for all clones. This
results in another bumpy matrix, except that now every cell of the
matrix is a character vector instead of a data frame.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tcr</span><span class="op">[</span>, , <span class="st">"Top V Hits"</span><span class="op">]</span></span>
<span><span class="co">#&gt; 2 x 5 BumpyCharacterMatrix</span></span>
<span><span class="co">#&gt; rownames: TRA TRB </span></span>
<span><span class="co">#&gt; colnames: sample1 sample2 sample3 sample4 sample5 </span></span>
<span><span class="co">#&gt; preview [1,1]:</span></span>
<span><span class="co">#&gt;   chr [1:90] "TRAV8-6, TRAV8-4" "TRAV26-1" "TRAV22" "TRAV6" "TRDV1" "TRAV17" ...</span></span></code></pre></div>
<p>Finally, and arguably most useful for us, it is also possible to
convert the bumpy matrix back into a long-form data frame. Here, the
column and rownames are added as additional columns to the data
frame:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BumpyMatrix</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BumpyMatrix/man/unsplitAsDataFrame.html" class="external-link">unsplitAsDataFrame</a></span><span class="op">(</span><span class="va">tcr</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/tibble/man/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">Clonal.Sequence</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 4,266 × 10</span></span></span>
<span><span class="co">#&gt;    row   column  Clone.Count Clone.Frequency Top.V.Hits       Top.J.Hits</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>         <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>           <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> TRA   sample1         193          0.055<span style="text-decoration: underline;">4</span> TRAV8-6, TRAV8-4 TRAJ5     </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> TRA   sample1         193          0.055<span style="text-decoration: underline;">4</span> TRAV26-1         TRAJ44    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> TRA   sample1         158          0.045<span style="text-decoration: underline;">3</span> TRAV22           TRAJ28    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> TRA   sample1         138          0.039<span style="text-decoration: underline;">6</span> TRAV6            TRAJ41    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> TRA   sample1         137          0.039<span style="text-decoration: underline;">3</span> TRDV1            TRAJ23    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> TRA   sample1         136          0.039<span style="text-decoration: underline;">0</span> TRAV17           TRAJ12    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> TRA   sample1         130          0.037<span style="text-decoration: underline;">3</span> TRAV13-1         TRAJ12    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> TRA   sample1         129          0.037<span style="text-decoration: underline;">0</span> TRAV19           TRAJ13    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> TRA   sample1         121          0.034<span style="text-decoration: underline;">7</span> TRAV13-1         TRAJ57    </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> TRA   sample1         118          0.033<span style="text-decoration: underline;">9</span> TRAV21           TRAJ12    </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 4,256 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 4 more variables: CDR3.Amino.Acid.Sequence &lt;chr&gt;, locus &lt;chr&gt;,</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;">#   sample &lt;chr&gt;, Top.D.Hits &lt;chr&gt;</span></span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Gregor Sturm.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
